### PR TITLE
feat: delete resources in batches of 10

### LINKED ISSUE #779: feat: Limit parallel deletes
### TL;DR

Limit the parallel ongoing deletes.

### Expected behavior

To avoid too much strain on the API and possible rate-limit exhaustions we should limit the number of parallel deletes (introduced in #761). This could be made configurable with a preference too.

Not sure what a good default limit would be. We could calculate something from the default API rate limit. In practice 5 is probably high enough to give some advantage to the user, while not doing too much in parallel.

### PR DESCRIPTION
**Pull request by jooola**
Closes #779 

- Delete the resources in batches of 10
- Reduce the complexity of the code by not using `goroutines`. We are already trying to limit the strain on the API by deleting in batches of 10, therefor waiting the actions of multiple delete calls is enough "concurrency" for this use case.

### CODE REVIEW

File: internal/cmd/base/delete.go [hunk 1]
```diff
@@ -53,48 +52,55 @@ func (dc *DeleteCmd) CobraCommand(s state.State) *cobra.Command {
...
 func (dc *DeleteCmd) Run(s state.State, cmd *cobra.Command, args []string) error {
+	errs := make([]error, 0, len(args))
+	deleted := make([]string, 0, len(args))
 
-	wg := sync.WaitGroup{}
-	wg.Add(len(args))
-	actions, errs :=
-		make([]*hcloud.Action, len(args)),
-		make([]error, len(args))
+	for _, batch := range util.Batches(args, 10) {
```
Comment (apricote) [inline internal/cmd/base/delete.go:58]: Can you move this magic number to a named variable/constant?

File: internal/cmd/base/delete.go [hunk 2]
```diff
@@ -53,48 +52,55 @@ func (dc *DeleteCmd) CobraCommand(s state.State) *cobra.Command {
...
+				results[i].Error = err
+				continue
+			}
+			if action != nil {
+				actions = append(actions, action)
+			}
+		}
+
+		if len(actions) > 0 {
+			// TODO: We do not check when an action fail for a specific resource
```
[the remark this answers is not in the document]
Comment (jooola) [reply internal/cmd/base/delete.go:89]: I do not see an easy way to solve this yet, I already had this issue when working on the fleeting plugin. We do not know from which resource object the actions originates.

Ideally, we should have a way to retrieve the resource ID and resource type from the action itself to be able to identify it.

The existing `[]ActionResource` might help, but once we have multiple values in the list, finding the right one feel like a dangerous game.

**Review by apricote — APPROVED** [review]
Approved from my side. Installed this locally and tested it to make sure it still works.

File: internal/cmd/util/resource.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package util
+
+type ResourceState struct {
```
[the remark this answers is not in the document]
Comment (jooola) [reply internal/cmd/util/resource.go:3]: I believe one can reuse this in other scenarios (work on multiple resources at once), so I moved it to the utils package.

I have a similar struct in the fleeting plugin, for the exact use case we have here. And we might at some point move our different implementations to a dedicated package in `hcloud-go`.

File: internal/cmd/base/delete.go [hunk 4]
```diff
@@ -51,50 +50,60 @@ func (dc *DeleteCmd) CobraCommand(s state.State) *cobra.Command {
 	return cmd
 }
 
+// DeleteBatchSize is the batch size when deleting multiple resources in parallel.
+const DeleteBatchSize = 10
```
[the remark this answers is not in the document]
Comment (jooola) [reply internal/cmd/base/delete.go:54]: For no particular reason :) Do you think I should make it private?

**Pull request merged**